code/modules/mob/living/carbon/give.dm

```diff
@@ -1,52 +1,164 @@
+	holder.mob.give_icon.icon_state = "act_give_off"
+	if(!silent)
+		to_chat(holder.mob, "<span class='info'>You're no longer trying to give someone your held item.</span>")
+	UnregisterSignal(holder.mob.get_active_hand(), list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED))
+	UnregisterSignal(holder.mob, list(SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), COMSIG_CARBON_SWAP_HANDS))
+	return ..()
+
+
+/datum/click_intercept/give/proc/cancel_give()
+	SIGNAL_HANDLER
+	qdel(src)
+
+
+/datum/click_intercept/give/InterceptClickOn(mob/user, params, atom/object)
+	if(user == object || !iscarbon(object))
+		return
+	var/mob/living/carbon/receiver = object
+	var/obj/item/I = user.get_active_hand()
+	if(!user.Adjacent(object))
+		to_chat(user, "<span class='warning'>You need to be closer to [receiver] to offer them [I]</span>")
+		return
+	if(!receiver.client)
+		to_chat(user, "<span class='warning'>You offer [I] to [receiver], but they don't seem to respond...</span>")
 		return
+	if((receiver.r_hand && receiver.l_hand) || HAS_TRAIT(receiver, TRAIT_HANDS_BLOCKED))
+		to_chat(user, "<span class='warning'>[receiver] doesn't have any hands free to accept [I]!</span>")
+		return
+	// We use UID() here so that the receiver can have more then one give request at one time.
+	// Otherwise, throwing a new "give item" alert would override any current one also named "give item".
+	receiver.throw_alert("give item [I.UID()]", /obj/screen/alert/take_item, alert_args = list(user, receiver, I))
+	to_chat(user, "<span class='info'>You offer [I] to [receiver].</span>")
+	qdel(src, FALSE, TRUE) // silent = TRUE so we don't give them the default chat message in Destroy.
+
+
+/**
+ * # Take Item alert
+ *
+ * Alert which appears for a user when another player is attempting to offer them an item.
+ * The user can click the alert to accept, or simply do nothing to not take the item.
+ */
+/obj/screen/alert/take_item
+	name = "Take Item"
+	desc = "someone wants to hand you an item!"
+	icon_state = "template"
+	timeout = 10 SECONDS
+	/// UID of the mob offering the receiver an item.
+	var/giver_UID
+	/// UID of the mob who has this alert.
+	var/receiver_UID
+	/// UID of the item being given.
+	var/item_UID
+
+
+/obj/screen/alert/take_item/Initialize(mapload, mob/giver, mob/receiver, obj/item/I)
+	. = ..()
+	desc = "[giver] wants to hand you \a [I]. Click here to accept it!"
+	giver_UID = giver.UID()
+	receiver_UID = receiver.UID()
+	item_UID = I.UID()
+	add_overlay(icon(I.icon, I.icon_state, SOUTH))
+	add_overlay("alert_flash")
+	ADD_TRAIT(giver, TRAIT_OFFERING_ITEM, TRAIT_GENERIC)
+	RegisterSignal(I, list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED), .proc/cancel_give)
+	RegisterSignal(giver, list(SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), COMSIG_CARBON_SWAP_HANDS), .proc/cancel_give)
+
+
+/obj/screen/alert/take_item/Destroy()
+	var/mob/living/giver = locateUID(giver_UID) // REMOVE_TRAIT needs a direct var reference.
+	REMOVE_TRAIT(giver, TRAIT_OFFERING_ITEM, TRAIT_GENERIC)
```

Comment:
do you need to check that `giver` is not null here? (I don't know)

Reply:
I think I'm going to negate the need for null checks by registering `COMSIG_PARENT_QDELETING` to the giver and receiver. That way if either starts to be deleted, the alert is deleted first, and we know they exist.